lib/ddtrace/contrib/faraday/middleware.rb

```diff
@@ -0,0 +1,52 @@
+require 'faraday'
+require 'ddtrace/ext/http'
+require 'ddtrace/ext/net'
+
+module Datadog
+  module Contrib
+    module Faraday
+      # Middleware implements a faraday-middleware for ddtrace instrumentation
+      class Middleware < ::Faraday::Middleware
+        def initialize(*)
+          @tracer = try_pin(:tracer, Datadog.tracer)
+          @service = try_pin(:service, SERVICE)
+
+          super
+        end
+
+        def call(env)
+          tracer.trace(SERVICE) do |span|
+            span.resource = env[:method].to_s.upcase
+            span.service = service
+            span.span_type = Ext::HTTP::TYPE
+            span.set_tag(Ext::HTTP::URL, env[:url].path)
+            span.set_tag(Ext::HTTP::METHOD, env[:method].upcase)
+            span.set_tag(Ext::NET::TARGET_HOST, env[:url].host)
+            span.set_tag(Ext::NET::TARGET_PORT, env[:url].port)
+            app.call(env).on_complete { |resp| handle_response(span, resp) }
+          end
+        end
+
+        private
+
+        attr_reader :app, :tracer, :service
+
+        def handle_response(span, env)
+          if Ext::HTTP::ERROR_RANGE.include?(env[:status])
+            span.set_error(["Error #{env[:status]}", env[:body]])
```

Comment:
not sure it's correct, a 404 may not be considered an error. Also users may have recovery policies in case of an issue (not 500), that may not be interpreted as an error.

I think if we provide something like that, it should be configurable. What if we set only 5xx as errors, and then discuss about this hot-topic that has been kept on-hold for so much time?

Reply:
I think we should maintain the same behavior as the `http` module (at least for now), which currently considers `4xx` status codes as errors. 

https://github.com/DataDog/dd-trace-rb/blob/master/lib/ddtrace/contrib/http/patcher.rb#L141-L146

About the recovery policy, I don't think that should change how an `500` status code is reported. On the HTTP level, an 500 is always an error. If the user handles that gracefully in the application-level, than an upper level span should reflect that.

What do you think?

Reply:
as a recovery policy I mean a service that you hit with a 404 (i.e. caching service) and that requires you to do a real query + write in the cache. It means having an error everytime you have a miss with a specific service. This is a specific use case, but was just to say that a 4xx is an error depending on the application itself.

Because `http` does that, we can keep that way for symmetry, but the point is we're not doing that in Python and only 5xx are considered errors. I think it's a wrong behavior in the `http` integration too (if it's not configurable).